### PR TITLE
Additional features around Suction Cups

### PR DESCRIPTION
**Pull request by tslater2006**
This contains a first pass repair algorithm for suction cups, as well as the ability to remove reported suction cups (and connected suction cup issues) from the list to prevent them from being repaired. Also contains a handy "contours intersect" helper method.

### CODE REVIEW

File: UVtools.Core/EmguCV/EmguContours.cs [hunk 1]
```diff
@@ -208,5 +209,32 @@ public static double[] GetContoursArea(List<VectorOfVectorOfPoint> contours, boo
             
             return result;
         }
+
+        public static bool CheckContoursIntersect(VectorOfVectorOfPoint contour1, VectorOfVectorOfPoint contour2)
+        {
+            var contour1Rect = CvInvoke.BoundingRectangle(contour1[0]);
+            var contour2Rect = CvInvoke.BoundingRectangle(contour2[0]);
+
```
Comment (sn4k3) [inline UVtools.Core/EmguCV/EmguContours.cs:217]: Before the hard code, check if both bouding rectangles intersect, if not return false and not worth to continue with code.

File: UVtools.Core/Layer/LayerManager.cs [hunk 2]
```diff
@@ -1493,7 +1493,22 @@ void CacheLayers(uint layerIndex, bool direction)
                             var suctionTrapArea = EmguContours.GetContourArea(trap);
                             if (suctionTrapArea < minimumSuctionArea) continue;
                             var rect = CvInvoke.BoundingRectangle(trap[0]);
-                            AddIssue(new LayerIssue(this[layerIndex], LayerIssue.IssueType.SuctionCup, trap.ToArrayOfArray(), rect) { Area = suctionTrapArea });
+                            var newIssue = new LayerIssue(this[layerIndex], LayerIssue.IssueType.SuctionCup, trap.ToArrayOfArray(), rect) { Area = suctionTrapArea };
+
+                            if (layerIndex < LayerCount - 1 && suctionTraps[layerIndex + 1] is not null)
```
Comment (sn4k3) [inline UVtools.Core/Layer/LayerManager.cs:1498]: Can this be moved to the repair method instead? and discard the ChildIssues and ParentIssue only used by this issue?

File: UVtools.Core/Operations/OperationRepairLayers.cs [hunk 3]
```diff
@@ -403,6 +406,64 @@ void initImage()
                 });
             }
 
+
+            if (_repairSuctionCups)
+            {
+                (bool canDrill, Point location) GetDrillLocation(LayerIssue issue, int diameter)
+                {
+                    var centroid = EmguCV.EmguContour.GetCentroid(new VectorOfPoint(issue.Contours[0]));
```
Comment (sn4k3) [inline UVtools.Core/Operations/OperationRepairLayers.cs:414]: centroid is not safe... If a suction is inside a ring like the bearing model, the centeroi  will be the center of the bearing, but suction is traped on the outter ring. 
Eg:
UVtools_2021-09-17_01-44-26

Centroid of hallow part is on center of the model, which will make a not valid position.
Maybe here we can use Distance transformation

Comment (tslater2006) [reply UVtools.Core/Operations/OperationRepairLayers.cs:414]: This is accounted for in the code by later ensuring that the centroid is actually in the contour area, this is achieved immediately below this where we draw the contour, and the planned drilling circle and ensure overlap. If there is no overlap we do not repair it.

File: UVtools.Core/EmguCV/EmguContours.cs (hunk 1 above)
Comment (tslater2006) [reply UVtools.Core/EmguCV/EmguContours.cs:217]: This has been addressed with a contour1Rect.IntersectsWith(contour2Rect) check and exiting if they do not.

File: UVtools.Core/Operations/OperationRepairLayers.cs (hunk 3 above)
Comment (sn4k3) [reply UVtools.Core/Operations/OperationRepairLayers.cs:414]: Using distance transformation and find the most itensity pixel will always ensure a center inside the issue, you can relate to centroid too. But can be a improvement for latter

UVtools_2021-09-17_01-56-31

File: UVtools.Core/Layer/LayerManager.cs (hunk 2 above)
Comment (tslater2006) [reply UVtools.Core/Layer/LayerManager.cs]: This should be addressed by most recent commit.

**Pull request merged**